### PR TITLE
Add “depositAndCall” functionality

### PR DESCRIPTION
**Pull request by gzeoneth**
- [x] Add “depositAndCall” functionality
- [x] Enable affordance for gateway to signal "depositAndCall" support
- [ ] Construct test cases for the new feature

### CODE REVIEW

File: packages/arb-bridge-peripherals/contracts/tokenbridge/arbitrum/gateway/L2ArbitrumGateway.sol [hunk 1]
```diff
@@ -243,14 +251,10 @@ abstract contract L2ArbitrumGateway is L2ArbitrumMessenger, TokenGateway {
...
+        (bytes memory gatewayData, bytes memory extraData) = GatewayMessageHandler
             .parseFromL1GatewayMsg(_data);
-
-        if (callHookData.length != 0) {
-            // callHookData should always be 0 since inboundEscrowAndCall is disabled
-            callHookData = bytes("");
-        }
-
+        // user encoded
+        (uint256 callHookGas, bytes memory callHookData) = abi.decode(extraData, (uint256, bytes));
```
Comment (fredlacs) [inline packages/arb-bridge-peripherals/contracts/tokenbridge/arbitrum/gateway/L2ArbitrumGateway.sol:257]: Let's instead handle this in the GatewayMessageHandler

**Review by fredlacs — CHANGES_REQUESTED** [review]
I think we should also add a pure function in the L1/L2 that signals that the gateway pair can handle the depositAndCall functionality (could be a pure function or erc165 style).
With this we can then make deposits revert on the L1 router if they have depositAndCall data, but are trying to use a token without that functionality

File: packages/arb-bridge-peripherals/contracts/tokenbridge/arbitrum/gateway/L2ArbitrumGateway.sol (hunk 1 above)
Comment (gzeoneth) [reply packages/arb-bridge-peripherals/contracts/tokenbridge/arbitrum/gateway/L2ArbitrumGateway.sol]: done in d8c24ac

**Comment by gzeoneth** [comment]
I think now the code should eliminated the griefing factors by
1. Ensure at some pt of execution we have at least callHookData.gas
2. Ensure the refund can be triggered by increasing gas limit

Actually we don't really need to check `gasleft() > callHookData.gas` if we assume there will not be gas dependent (not oog) code in the call hook, if someone execute with a low gas limit it will oog and revert and the retryable ticket can be executed again.

We might need something to prevent user setting callHookData.gas too large that they cannot pay for it on L2, but I don't want to do decode on L1 due to gas cost. We can use uint24 so they cannot set it to more than 16m gas and with encodePacked we can save some calldata gas packing with the address field, or we can hard code the limit L2. 

Also note that currently `callHookData.gas` is not changed on L1 tradable exit.

**Comment by fredlacs** [comment]
the above avoid a potential error of the abi encoding of the user data failing.
we instead pass the refundAddress and maxGas as parameters to a new entrypoint function

**Comment by Jason-Wanxt** [comment]
We should encode token's address to callHookData because in standard transferAndCall, the target contract can call msg.sender to get the token's address, but for our implement, it's hard for the target contract to get that address.

**Comment by gzeoneth** [comment]
> We should encode token's address to callHookData because in standard transferAndCall, the target contract can call msg.sender to get the token's address, but for our implement, it's hard for the target contract to get that address.

That's true but I think it might be better to do in tooling?

**Comment by fredlacs** [comment]
I think it's best if we add it to the onTokenTransfer function signature instead

**Comment by Jason-Wanxt** [comment]
Yea, but if it will be better if we add a check on inboundEscrowAndCall to avoid user encode a wrong l2 token address for any reasons?

**Comment by fredlacs** [comment]
should we forward msg.value on the `onTokenTransfer`

**Comment by gzeoneth** [comment]
> should we forward msg.value on the `onTokenTransfer`

there shouldn't be any callvalue as is
https://github.com/OffchainLabs/arbitrum/blob/03bb55b5c1936d0838baa639b020d291230beb9c/packages/arb-bridge-peripherals/contracts/tokenbridge/ethereum/gateway/L1ArbitrumGateway.sol#L206

**Comment by fredlacs** [comment]
we should also assess issues brought up in https://github.com/OffchainLabs/arbitrum/issues/2255

**Pull request closed without merging**